packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,434 @@
+    }
+    return harden(encodeToSmallcapsRecur(passable));
+  };
+  return harden(encodeToSmallcaps);
+};
+harden(makeEncodeToSmallcaps);
+
+/**
+ * @typedef {object} DecodeFromSmallcapsOptions
+ * @property {(encodedRemotable: SmallcapsEncoding
+ * ) => Remotable} [decodeRemotableFromSmallcaps]
+ * @property {(encodedPromise: SmallcapsEncoding
+ * ) => Promise} [decodePromiseFromSmallcaps]
+ * @property {(encodedError: SmallcapsEncoding
+ * ) => Error} [decodeErrorFromSmallcaps]
+ */
+
+const dontDecodeRemotableFromSmallcaps = encoding =>
+  assert.fail(X`remotable unexpected: ${encoding}`);
+const dontDecodePromiseFromSmallcaps = encoding =>
+  assert.fail(X`promise unexpected: ${encoding}`);
+const dontDecodeErrorFromSmallcaps = encoding =>
+  assert.fail(X`error unexpected: ${q(encoding)}`);
+
+/**
+ * @param {DecodeFromSmallcapsOptions=} decodeOptions
+ * @returns {(encoded: SmallcapsEncoding) => Passable}
+ */
+export const makeDecodeFromSmallcaps = ({
+  decodeRemotableFromSmallcaps = dontDecodeRemotableFromSmallcaps,
+  decodePromiseFromSmallcaps = dontDecodePromiseFromSmallcaps,
+  decodeErrorFromSmallcaps = dontDecodeErrorFromSmallcaps,
+} = {}) => {
+  /**
+   * `decodeFromSmallcaps` may rely on `encoding` being the result of a
+   * plain call to JSON.parse. However, it *cannot* rely on `encoding`
+   * having been produced by JSON.stringify on the output of `encodeToSmallcaps`
+   * above, i.e., `decodeFromSmallcaps` cannot rely on `encoding` being a
+   * valid marshalled representation. Rather, `decodeFromSmallcaps` must
+   * validate that.
+   *
+   * @param {SmallcapsEncoding} encoding must be hardened
+   */
+  const decodeFromSmallcaps = encoding => {
+    switch (typeof encoding) {
+      case 'boolean':
+      case 'number': {
+        return encoding;
+      }
+      case 'string': {
+        if (encoding === '') {
+          return encoding;
+        }
+        switch (encoding[0]) {
+          case "'": {
+            // un-hilbert-ify the string
+            return encoding.slice(1);
+          }
+          case '%': {
+            return passableSymbolForName(encoding.slice(1));
+          }
+          case '#': {
+            switch (encoding) {
+              case '#undefined': {
+                return undefined;
+              }
+              case '#NaN': {
+                return NaN;
+              }
+              case '#Infinity': {
+                return Infinity;
+              }
+              case '#-Infinity': {
+                return -Infinity;
+              }
+              default: {
+                assert.fail(X`unknown constant "${q(encoding)}"`, TypeError);
+              }
+            }
+          }
+          case '+':
+          case '-': {
+            const last = encoding.length - 1;
+            assert(
+              encoding[last] === 'n',
+              X`Encoded bigint must start with "+" or "-" and end with "n": ${encoding}`,
+            );
```

Comment:
I believe @dtribble suggested dropping the trailing `n`, which would allow us to avoid the `slice` call,

Reply:
Done